vscode/src/services/telemetryV2.ts

```
@@ -0,0 +1,56 @@
+import { ConfigurationWithAccessToken } from '@sourcegraph/cody-shared/src/configuration'
+import {
+    BillingCategory,
+    BillingProduct,
+    EventAction,
+    EventFeature,
+    MetadataKey,
+} from '@sourcegraph/cody-shared/src/telemetry-v2'
+import { TelemetryRecorderProvider } from '@sourcegraph/cody-shared/src/telemetry-v2/TelemetryRecorderProvider'
+import { TelemetryRecorder } from '@sourcegraph/telemetry'
+
+import { localStorage } from './LocalStorageProvider'
+import { extensionDetails } from './telemetry'
+
+let telemetryRecorderProvider: TelemetryRecorderProvider | undefined
+
+/**
+ * Get a recorder for recording telemetry events in the new telemetry framework:
+ * https://docs.sourcegraph.com/dev/background-information/telemetry
+ *
+ * A new recorder should generally be retrieved in order to pick up configuration
+ * changes.
+ */
+export function getRecorder():
+    | TelemetryRecorder<EventFeature, EventAction, MetadataKey, BillingCategory, BillingProduct>
+    | undefined {
+    return telemetryRecorderProvider?.getRecorder()
+}
+
+export async function createOrUpdateTelemetryRecorderProvider(
+    config: ConfigurationWithAccessToken,
+    isExtensionModeDevOrTest: boolean
+): Promise<void> {
+    if (config.telemetryLevel === 'off' || isExtensionModeDevOrTest) {
```

It's helpful to log events into the console if `isExtensionModeDevOrTest`. That's what we do with the current `logEvent` implementation. It can be implemented as a wrapper around the recorder-getter with the following API:

```ts
import { recordEvent } from '...'

// If `isExtensionModeDevOrTest === true` logs event to the console.
// No need to call a getter function with a potentially `undefined` result.
recordEvent('cody', 'installed')
```

We can throw an error if the underlying `telemetryRecorderProvider` is not initialized because we always want it to be in place. We use this pattern for other singleton entities (e.g., `localStorage`).

---

Added a console exporter and error throwing if used before init!